### PR TITLE
Impro1816 handle realization coord for prever system

### PR DESCRIPTION
**Pull request by MoseleyS**
Adds a blend-cycles-and-realizations CLI

Adds an option to collapse the realization coord at the start of the spot-extract CLI

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
- improver/cli/blend_cycles_and_realizations.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/blend_cycles_and_realizations.py  2
- improver/cli/spot_extract.py  1
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
- improver/cli/blend_cycles_and_realizations.py  2
- improver/cli/spot_extract.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/spot_extract.py  1
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
- improver/cli/blend_cycles_and_realizations.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

File: improver/cli/blend_cycles_and_realizations.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+from improver import cli
+
+
+@cli.clizefy
+@cli.with_output
+def process(
+    *cubes: cli.inputcube, cycletime: str = None,
+):
+    """Runs equal-weighted blending.
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/blend_cycles_and_realizations.py]: I've expanded this a little.

File: improver/cli/blend_cycles_and_realizations.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    forecast_reference_time coordinates.
+
+    Args:
+        cubes (iris.cube.CubeList):
+            Cubelist of cubes to be blended.
+        cycletime (str):
+            The forecast reference time to be used after blending has been
+            applied, in the format YYYYMMDDTHHMMZ. If not provided, the
+            blended file takes the latest available forecast reference time
+            from the input datasets supplied.
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/blend_cycles_and_realizations.py]: Supplied removed

File: improver/cli/blend_cycles_and_realizations.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    Returns:
+        iris.cube.Cube:
+            Merged and blended Cube.
+    """
+    from improver.blending.calculate_weights_and_blend import WeightAndBlend
+    from iris.cube import CubeList
+
+    plugin = WeightAndBlend("realization", "linear", y0val=0.5, ynval=0.5,)
+    cubelist = CubeList([])
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/blend_cycles_and_realizations.py]: This is just me. Square-brackets removed.

File: improver/cli/spot_extract.py [hunk 4]
```diff
@@ -87,7 +88,8 @@ def process(
             data corresponding to those percentiles will be returned. For
             example "25, 50, 75" will result in the 25th, 50th and 75th
             percentiles being returned from a cube of probabilities,
-            percentiles or realizations. Deterministic input data will raise
+            percentiles or realizations. If a realization coord is found, it will be
+            collapsed using equal-weights. Deterministic input data will raise
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/spot_extract.py]: I've removed the added sentence as I don't think it is needed. I suspect I had a plan to do this and then found it necessary to add a new argument instead.

File: improver/cli/spot_extract.py [hunk 5]
```diff
@@ -102,6 +104,8 @@ def process(
         suppress_warnings (bool):
             Suppress warning output. This option should only be used if it
             is known that warnings will be generated but they are not required.
+        realization_collapse (bool):
+            Triggers equal-weighting blending of the realization coord if required.
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/spot_extract.py:107]: Expanded

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/spot_extract.py  1
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
- improver/cli/blend_cycles_and_realizations.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

**Review by bayliffe — APPROVED** [review]

File: improver/cli/spot_extract.py [hunk 6]
```diff
@@ -149,6 +154,8 @@ def process(
     from improver.spotdata.spot_extraction import SpotExtraction
     from improver.utilities.cube_extraction import extract_subcube
 
+    if realization_collapse:
+        cube = WeightAndBlend("realization", "linear", y0val=0.5, ynval=0.5,)(cube)
```
Comment (cgsandford) [inline improver/cli/spot_extract.py:158]: Why are we always using this and not:
```
cube = collapsed(cube, "realization", iris.analysis.MEAN)
cube.remove_coord("realization")
```
?

File: improver/cli/blend_cycles_and_realizations.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        iris.cube.Cube:
+            Merged and blended Cube.
+    """
+    from improver.blending.calculate_weights_and_blend import WeightAndBlend
+    from iris.cube import CubeList
+
+    plugin = WeightAndBlend("realization", "linear", y0val=0.5, ynval=0.5,)
+    cubelist = CubeList()
+    for cube in cubes:
+        cubelist.append(plugin(cube))
```
Comment (cgsandford) [inline improver/cli/blend_cycles_and_realizations.py:68]: Again, we can use:
```
cubelist = CubeList()
for cube in cubes:
    cube = collapsed(cube, "realization", iris.analysis.MEAN)
    cube.remove_coord("realization")
    cubelist.append(cube)
```
Honestly, realization collapse is done so often and is so simple it is worth just writing a little function in `improver.utilities.cube_manipulation` to do this.

**Review by cgsandford — CHANGES_REQUESTED** [review]
Preference not to call a very complicated plugin where we don't have to.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/blend_cycles_and_realizations.py  2
- improver/cli/spot_extract.py  1
- improver_tests/utilities/cube_manipulation/test_collapse_realizations.py  3
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

File: improver/cli/blend_cycles_and_realizations.py (hunk 7 above)
Comment (MoseleyS) [reply improver/cli/blend_cycles_and_realizations.py]: Ok. I've accepted this recommentation and added a small function (with test) to do this.

File: improver/cli/spot_extract.py (hunk 6 above)
Comment (MoseleyS) [reply improver/cli/spot_extract.py]: Changed.

**Review by cgsandford — DISMISSED** [review]
Code is good - just running acceptance tests and will then approve the ticket.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapse_realizations.py  3
- improver_tests/acceptance/test_blend_cycles_and_realizations.py  1
- improver/cli/blend_cycles_and_realizations.py  2
- improver/cli/spot_extract.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6004184&bid=19683074)

**Review by cgsandford — APPROVED** [review]

**Pull request merged**